central/pruning/pruning.go

```diff
@@ -846,31 +856,64 @@ func (g *garbageCollectorImpl) collectAlerts(config *storage.PrivateConfig) {
 			AddExactMatches(search.ViolationState, storage.ViolationState_ATTEMPTED.String()).
 			AddDays(search.ViolationTime, int64(pruneAttemptedRuntimeAfter)).
 			ProtoQuery()
-		queries = append(queries, q)
+		queryMap[pruneAttemptedRuntimeAfterKey] = q
 	}
 
-	if len(queries) == 0 {
+	if len(queryMap) == 0 {
 		log.Info("No alert retention configuration, skipping")
 		return
 	}
 
 	ctx, cancel := context.WithTimeout(pruningCtx, alertQueryTimeout)
 	defer cancel()
 
-	alertResults, err := g.alerts.Search(ctx, search.DisjunctionQuery(queries...))
-	if err != nil {
-		log.Error(err)
-		return
-	}
+	prunedAlertCount := 0
+	// Go through the various queries and remove those batches.
+	for key, query := range queryMap {
+		log.Infof("Pruning for key %q", key)
+		alertResults, err := g.alerts.Search(ctx, query)
+		if err != nil {
+			log.Error(err)
+			return
+		}
 
-	alertsToPrune := search.ResultsToIDs(alertResults)
+		alertsToPrune := make([]string, 0, len(alertResults))
+		// The alert searcher is opinionated and adds some default query parameters.
+		// Those should not be included for pruning.  Simpler to just use `WalkByQuery`
+		err = g.alerts.WalkByQuery(ctx, query, func(alert *storage.Alert) error {
+			alertsToPrune = append(alertsToPrune, alert.GetId())
+			return nil
+		})
+		if err != nil {
+			log.Errorf("Unable to prune alerts for query %v", query)
+			continue
+		}
 
-	if len(alertsToPrune) > 0 {
-		log.Infof("[Alert pruning] Removing %d alerts", len(alertsToPrune))
-		if err := g.alerts.DeleteAlerts(pruningCtx, alertsToPrune...); err != nil {
-			log.Error(err)
+		localBatchSize := alertDeleteBatchSize
+
+		for {
+			if len(alertsToPrune) == 0 {
+				break
+			}
+
+			if len(alertsToPrune) < localBatchSize {
+				localBatchSize = len(alertsToPrune)
+			}
+
+			alertBatch := alertsToPrune[:localBatchSize]
+			log.Infof("[Alert pruning] Removing %d alerts for %q", len(alertBatch), key)
+			if err := g.alerts.DeleteAlerts(pruningCtx, alertBatch...); err != nil {
+				log.Error(err)
+			} else {
+				prunedAlertCount = prunedAlertCount + len(alertBatch)
```

Comment:
Why is the prunedAlertCount increased when `DeleteAlerts` returns an error? I would intentionally think that alerts weren't pruned.

Reply:
I'm not following.  That is in the else as in err is nil.

Reply:
Ah, I've not read it correctly. The logic looks correct.